Allow opt-out from client certificate root checking

edwardneal opened the pull request:
Fixes #14718. 

This aligns the code with README.md, and enables chained certificate authentication to work in Azure App Service (which won't have the client certificate's root in its trust store.)

edwardneal committed:
Added a new property, ValidateCertificateChain

This aligns the code with README.md, and enables chained certificate authentication to work in Azure App Service (which won't have the client certificate's root in its trust store)

edwardneal committed:
Updated reference assemblies

New ValidateCertificateChain property was missing from /ref - hadn't updated it, so automatic test was failing

anurse commented:
Seems reasonable to me. @blowdart any reason we did these bools instead of just exposing the verification flags themselves?

blowdart commented:
Simplicity. Users shouldn't have to know the flags when true/false works just as well. 

I'd rather see the other approach of wiring up the custom root store (https://github.com/dotnet/corefx/pull/40379) it's more flexible in this regard rather than saying "Allow anything"

anurse commented:
> Users shouldn't have to know the flags when true/false works just as well.

I don't really buy that when flags are used in a whole bunch of places, and this creates a need for new documentation and usage patterns rather than just reusing existing paradigms/docs/etc., but whatever. The existing `Validate` bools exist and I'm not going to advocate for a wholesale change here.

> I'd rather see the other approach of wiring up the custom root store ([dotnet/corefx#40379](https://github.com/dotnet/corefx/pull/40379)) it's more flexible in this regard rather than saying "Allow anything"

@edwardneal can you adjust this PR to follow this suggestion? Basically it sounds like we'd expose an `X509Certificate2Collection` on our options to use as the `CustomTrustStore` value added to `X509Chain` in 5.0.

edwardneal commented:
Sure.

If I replace the `ValidateCertificateChain` property with a custom `X509Certificate2Collection`, it won't be possible to bypass the issuer check, just specify a list of known-good issuers. That'll be a bit more secure by default and I can't think of a production scenario where we're authenticating against issuers we don't know about. Are there any corner cases where that capability might be useful enough to justify keeping the other property as well?

edwardneal committed:
Replaced the broad property which disables certificate chain validation with a granular list of trusted issuers

Also updated documentation to remove the unused property and to indicate that it isn't possible to validate half of the chain from the OS' trust store, and half from the private root store

edwardneal committed:
Swapped property access for an AddRange call

edwardneal committed:
Updated reference assembly

anurse commented:
Can you add a test? It should be feasible to just grab one of our existing tests and augment it to test this scenario. If necessary, it's OK to check in a `.cer` file with the public side of a self-signed cert (let's avoid checking in a `.pfx` even though it's a self-signed test cert to avoid getting people worried ;)).

davidsh commented:
>If necessary, it's OK to check in a .cer file with the public side of a self-signed cert (let's avoid checking in a .pfx even though it's a self-signed test cert to avoid getting people worried ;)).

.NET Core has APIs to generate certificates with private keys including self-signed or chain-signed etc. So, you can avoid checking in *.cer or *.pfx files. Instead, just have the test generate what you need directly.

edwardneal committed:
Added certificate tests

Three test cases:
- Untrusted client certificate should fail, as it does now.
- Client certificate which is part of a chain with an untrusted root but a trusted issuer should fail, since the chain as a whole isn't trusted.
- Client certificate which is part of a fully-trusted chain should pass.

Also noticed that there were a couple of issues in CreateServer, where the certificate revocation flags weren't being set properly.

Finally, the csproj was referring to a folder name ("test\Certificates") which didn't exist. It looks like the folder was renamed to "TestCertificates", so I've update the project to match.

edwardneal commented:
I've added three tests: one to make sure the old behaviour is still working, one to test the new behaviour, and one which deals with a more complex scenario where there's an incomplete certificate chain (i.e. the client certificate's issuer is trusted, but the root of that chain is not.)

HaoK committed:
Merge branch 'master' into control-client-certificate-root-validation

HaoK committed:
Update Microsoft.AspNetCore.Authentication.Certificate.netcoreapp.cs

HaoK commented:
@edwardneal looks like there's a bunch of test failures, are you sure you want to rename the certificate directories as part of this?

edwardneal commented:
@HaoK no, definitely not. I noticed that my new certificates weren't copying from the TestCertificates directory to the output directory and saw that the project file was referring to `test\Certificates`. I thought that `Certificates` had been renamed to `TestCertificates` at some point, and updated the project file to match. I'll change the path back.

edwardneal committed:
Reverting change of Tests csproj, which was blocking certificate tests

edwardneal committed:
Forgot to add certificates to the shared source root, so tests were failing

edwardneal commented:
It looks like one of the tests isn't passing: `VerifyValidClientCertWithTrustedChainAuthenticates`. This is only happening on MacOS though; on the last build, it passed on Windows and Linux.

I don't think it's an ASP.NET-specific problem: the only new functionality which the PR uses is the custom trust store, so it feels like it's an issue in the underlying framework.

I've not got access to a Mac to test it out, so I can't really isolate or resolve the problem. How do you want to carry on here? I could raise a bug and we could leave the PR open until it's fixed, or we could merge the PR and raise a bug in parallel.

HaoK commented:
Can't merge until it passes, hopefully it passes on rerun

HaoK commented:
@edwardneal looks like there are real failures on mac right now from your new tests

```
   at Microsoft.AspNetCore.Authentication.Certificate.Test.ClientCertificateAuthenticationTests.VerifyValidClientCertWithTrustedChainAuthenticates() in /_/src/Security/Authentication/test/CertificateTests.cs:line 298
--- End of stack trace from previous location ---
```

edwardneal commented:
Thanks for the heads up. I was hoping it'd just be transient!

I've just tried to find out if the cross-platform components of the underlying X509Chain functionality are working by looking for the results of the [X509Chain tests](https://github.com/dotnet/corefx/blob/master/src/System.Security.Cryptography.X509Certificates/tests/ChainTests.cs) but can't find them. The Pipelines build is automatically passing without recording any results for that platform, and I can't see anything relevant for the dotnet user in mc.dot.net. It looks to me like tests aren't being run. Do you know if there's anywhere else that the automatic test results might be stored?

I'll put a simple piece of test code together and raise a bug in corefx.

HaoK commented:
@edwardneal looks like on mac we always get forbidden from all 3 tests you added, it doesn't look like there's anything mac specific in the code here, so maybe we are hitting an issue with the underlying X509 apis

I'm not sure if this is the same issue we are hitting, @blowdart does this sound like it might be the same thing?  https://github.com/dotnet/corefx/issues/29996

HaoK commented:
This is another possible related client cert bug https://github.com/dotnet/corefx/issues/34740

HaoK commented:
FYI @scottaddie we'll need to update the Cert Auth docs with this feature once its in

```
### ValidateCertificateChain
### CustomTrustStore

This check validates that the issuer for the certificate is trusted by the application host OS. If you are going to accept self-signed certificates you must disable this check.
This check validates that the certificate presented by the client can form a valid certificate chain when combined with the certificates in this collection.

If any certificates are added to this collection then the default OS trust store is not used to validate the certificate chain, so this collection needs to contain the complete certificate chain.
```

I'll resolve the conflict in readme which will lose this change

HaoK committed:
Merge branch 'master' into control-client-certificate-root-validation

edwardneal commented:
That's the conclusion I was coming to too. The only new API call it triggers (as far as I can tell) is the reference to `CustomTrustStore`, so it feels like there's *something* going on in there. I just can't debug it to find out what that *something* is.

blowdart commented:
@bartonjs will know for sure and should review the PR too

bartonjs commented on src/Security/Authentication/Certificate/ref/Microsoft.AspNetCore.Authentication.Certificate.netcoreapp.cs:
```diff
@@ -24,6 +24,7 @@ public partial class CertificateAuthenticationOptions : Microsoft.AspNetCore.Aut
     {
         public CertificateAuthenticationOptions() { }
         public Microsoft.AspNetCore.Authentication.Certificate.CertificateTypes AllowedCertificateTypes { [System.Runtime.CompilerServices.CompilerGeneratedAttribute] get { throw null; } [System.Runtime.CompilerServices.CompilerGeneratedAttribute] set { } }
+        public System.Security.Cryptography.X509Certificates.X509Certificate2Collection CustomTrustStore { [System.Runtime.CompilerServices.CompilerGeneratedAttribute] get { throw null; } [System.Runtime.CompilerServices.CompilerGeneratedAttribute] set { } }
```
In X509Chain we made it explicit that you had to set the TrustMode.  (Just adding something to the collection will fail, since you're in conflicting modes).

I recommend being explicit here, too.

edwardneal replied:
I always set TrustMode to `CustomRootTrust` when the trust store contains certificates - logic for this is [here](https://github.com/aspnet/AspNetCore/blob/1d9d78d4be5ec33191b46a9361fcc342a0a1cc7c/src/Security/Authentication/Certificate/src/CertificateAuthenticationHandler.cs#L170).

I'll expose the existing TrustMode property from CertificateAuthenticationOptions, unless you'd prefer that I use a separate enum and convert the values.

bartonjs replied:
The same enum is fine with me.  The concern is that if/when we add more modes this code can't adapt if it implicitly sets the mode by there being certs in this collection.

So please don't do anything implicit.

bartonjs commented:
I don't see anything that would cause it to fail; but I can't easily see details about the certs.  Maybe the test root has some characteristic macOS doesn't like.

Using cert chains generated by test process also solves the problem of certs eventually expiring.

edwardneal committed:
Updating following code review

Replaced custom logic with a new ChainTrustValidationMode property. Updated reference assemblies

edwardneal committed:
Changed tests to account for new property

HaoK commented:
Cert info:
```
SelfSignedPrimaryRoot:
Issuer: Valid Self Signed Primary Root
Valid from: Monday, Nov 4, 2019
Valid to: Monday, Dec 31, 2068
Subject: Valid Self Signed Primary Root
Public key parameters: 05 00
Enhanced Key Usage:
Server Authentication (1.3.6.1.5.5.7.3.1)
Client Authentication (1.3.6.1.5.5.7.3.2)
Time Stamping (1.3.6.1.5.5.7.3.8)
OCSP Signing (1.3.6.1.5.5.7.3.9)

Key Usage: Certificate Signing, Off-line CRL Signing, CRL Signing (06)
Basic Constraints: 
Subject Type=CA
Path Length Constraint=None




```

HaoK commented:
Anything look weird in the cert info for mac OS @bartonjs ?

anurse commented:
Ping @bartonjs (now that the holidays are over ;)).

bartonjs commented:
This is probably failing claiming that the hostname doesn't match, but that's just because of extra tests that Apple added in the pipeline that we use to determine hostname validity (https://support.apple.com/en-us/HT210176).

Your "signed client" cert is good for 50 years, which exceeds their 825 day limit Apple applies to TLS Server certs in 10.13; and so when we ask "does this cert meet the TLS policy for this hostname" (which used to just be "is the hostname right?") they say "no".

If it's actually having a problem with the root certs, it is unusual to have EKU on a root cert, and the test certs that I generate as part of my dynamic chain tests for corefx also assert DigitalSignature in their KeyUsage.

edwardneal committed:
Updated certificates for PR tests

edwardneal commented:
Thanks. I've regenerated the certificates, but whatever's going wrong doesn't seem to be related to most of those checks as far as I can tell. I've built the chain with identical durations and similar (E)KUs to the GitHub certificate chain, which are probably fine. The only difference between the two chains is that ValidSelfSignedPrimaryRootCertificate doesn't have Secure Email, Code Signing and Time Stamping EKUs, but I'm not using any of that functionality.

The client certificate doesn't have the computer name as a sAN, but I figure that should be okay - we're trying to identify a user, not the server.

```
ValidSelfSignedPrimaryRootCertificate:
Issuer: Valid Self Signed Primary Root
Valid from: Saturday, Jan 11, 2020
Valid to: Wednesday, Jan 11, 2045
Subject: CN = Valid Self Signed Primary Root
Public key parameters: 05 00
Enhanced Key Usage:
Server Authentication (1.3.6.1.5.5.7.3.1)
Client Authentication (1.3.6.1.5.5.7.3.2)

Key Usage: Digital Signature, Certificate Signing, Off-line CRL Signing, CRL Signing (86)
Basic Constraints: 
Subject Type=CA
Path Length Constraint=None


ValidSignedSecondaryRootCertificate:
Issuer: CN = Valid Self Signed Primary Root
Valid from: Saturday, Jan 11, 2020
Valid to: Thursday, Jan 11, 2035
Subject: Valid Signed Secondary Root
Public key parameters: 05 00
Enhanced Key Usage:
Server Authentication (1.3.6.1.5.5.7.3.1)
Client Authentication (1.3.6.1.5.5.7.3.2)

Key Usage: Digital Signature, Certificate Signing, Off-line CRL Signing, CRL Signing (86)
Basic Constraints: 
Subject Type=CA
Path Length Constraint=None


ValidSignedClientCertificate:
Issuer: CN = Valid Signed Secondary Root
Valid from: Saturday, Jan 11, 2020
Valid to: Tuesday, Jan 11, 2022
Subject: Valid Signed Client
Public key parameters: 05 00
Enhanced Key Usage:
Server Authentication (1.3.6.1.5.5.7.3.1)
Client Authentication (1.3.6.1.5.5.7.3.2)

Key Usage: Digital Signature, Key Encipherment (a0)
Basic Constraints: 
Subject Type=End Entity
Path Length Constraint=None

```

bartonjs commented:
Without seeing what errors you're actually getting in the X509Chain result, it's hard to diagnose further.

edwardneal commented:
I was able to run the test on a Mac, and have narrowed it down slightly. It isn't as helpful as I was hoping though.

The only response from X509Chain is `PartialChain`, so it seems to me that the certificate key usages and EKUs are correct and the problem just lies in being able to chain them together. MacOS can't process the certificate chain either though: running the command:

`security verify-cert -c validSignedClientCertificate.cer -r validSignedSecondaryRootCertificate.cer -r validSelfSignedPrimaryRootCertificate.cer -v -L`

returns the output:

```
Cert Verify Result: CSSMERR_TP_NOT_TRUSTED
---
Certificate chain
 0: Valid Sign Client
    <cert(0x7fcf57c07d20) s: Valid Signed Client i: Valid Signed Secondary Root>
---
Certificate errors
 0: Valid Signed Client
    Unable to find next certificate in the chain [MissingIntermediate]
---
Trust evaluation results
{
    TrustEvaluationDate = "2020-01-16 23:58:23 +0000";
    TrustResultDetails =      (
                {
            MissingIntermediate = 0;
            StatusCodes =                (
                "-2147409622"
            );
        }
    );
    TrustResultValue = 5;
}
---
Trust evaluation errors
Error Domain=NSOSStatusErrorDomain Code=-25318 "errKCCreateChainFailed / errSecCreateChainFailed: / The attempt to create a certificate chain failed." UserInfo={NSLocalizedDescription=\u201cValid Signed Client\u201d certificate is not trusted, NSUnderlyingError=0x7fcf57d004a0 {Error Domain=NSOSStatusErrorDomain Code=-25318 "errKCCreateChainFailed / errSecCreateChainFailed: / The attempt to create a certificate chain failed."
UserInfo={NSLocalizedDescription=Certificate 0 \u201cValid Signed Client\201d has errors: Unable to build chain to root (possible missing intermediate);}}}
```

I can't see why it'd fail to find the issuer: the Issuer of the client certificate is byte-for-byte identical to the Subject of its issuing certificate. Interestingly, I see the same problem if I add the root and second-level certificates to the system's keychain and mark them as trusted.

I'm planning to generate certificates from MacOS and try to validate them, but this doesn't seem to be coming from ASP.NET or the .NET Framework.

edwardneal commented:
Adding some more test results.

**openssl verify**:
```
> openssl verify -verbose -CAfile validSignedCertificates.pem -CApath . validSignedClientCertificate.pem
validSignedClientCertificate.pem: OK
````
I converted the three certificates we care about from DER-encoded .cer files to .pem files, then concatenated validSelfSignedPrimaryRootCertificate.pem and validSignedSecondaryRootCertificate.pem together to form validSignedCertificate.pem.

**security verify-cert**:
```
> security verify-cert -c validSignedClientCertificate.cer -r validSignedClientCertificate.cer -r validSignedSecondaryRootCertificate.cer -r validSelfSignedPrimaryRootCertificate.cer -v -L
...certificate verification successful.
---
Certificate chain
 0: Valid Signed Client
    <cert(0x7f8c14407db0) s: Valid Signed Client i: Valid Signed Secondary Root
---
Trust evaluation results
{
    TrustEvaluationDate = "2020-01-18 21:11:40 +0000";
    TrustResultDetails =     {
                {
        }
    );
    TrustResultValue = 4;
}
```
I don't think the second test result is worth much: I added `-r validSignedClientCertificate.cer` because the documentation for `SecTrustCreateWithCertificates` says that the certificate list includes the original certificate and not just its issuers. It still succeeds even if I remove `-r validSignedSecondaryRootCertificate.cer -r validSelfSignedPrimaryRootCertificate.pem` from the command though, so I'm pretty sure that this has just bypassed the validation checks entirely - not the direction I'm aiming for.

I've got a small test case which excludes ASP.NET and just references `X509Chain`, so I'll raise a bug over at dotnet/runtime tomorrow.

edwardneal committed:
Updated certificates

Updating all new certificates to avoid using the alternate signature algorithm

anurse commented:
Looks like there are build failures now: 

```
##[error]CertificateTests.cs(277,30): error CS0103: The name 'sucessfulValidationEvents' does not exist in the current context
```

edwardneal commented:
Sorry about the delay, I've been on the road for a bit. I don't know where that error's come from - updating the certificates shouldn't have caused a compilation error, I can perform a clean build and test run from my end, and there are a good dozen other nearly-identical test methods which reference the same variable and seem to compile well enough.

I'll perform a dummy checkin to see if there was some kind of transient problem with that particular build.

edwardneal committed:
Dummy checkin

Compilation errors coming from CertificateTests.cs, performing this checkin to test whether or not they're limited to a specific build.

edwardneal committed:
Merge branch 'master' into control-client-certificate-root-validation

edwardneal committed:
Following up typo corrections

Commit 116799fa709ff003781368b578e4efe2fa32e937 in upstream repo changed a variable name, causing compile errors on PR build. Fixing that

edwardneal commented:
I've pulled the current state of the repo down into my fork and found the problem - commit 116799fa709ff003781368b578e4efe2fa32e937 renamed a variable to correct a typo, and I hadn't accounted for that. The build's now looking a lot more promising.

HaoK commented:
Nice @edwardneal thanks for sticking with this and figuring out the issue with the mac certs, this pr looks great!
